pkg/policy/api/rule_validation_test.go

```diff
@@ -237,6 +237,33 @@ func (s *PolicyAPITestSuite) TestL7RulesWithNonTCPProtocols(c *C) {
 
 }
 
+// This test ensures that L7 rules reject unspecified ports.
+func (s *PolicyAPITestSuite) TestL7RuleRejectsEmptyPort(c *C) {
+	invalidL7PortRule := Rule{
+		EndpointSelector: WildcardEndpointSelector,
+		Ingress: []IngressRule{
+			{
+				IngressCommonRule: IngressCommonRule{
+					FromEndpoints: []EndpointSelector{WildcardEndpointSelector},
+				},
+				ToPorts: []PortRule{{
+					Ports: []PortProtocol{
+						{Port: "0", Protocol: ProtoTCP},
+					},
+					Rules: &L7Rules{
+						HTTP: []PortRuleHTTP{
+							{},
+						},
+					},
+				}},
+			},
+		},
+	}
+
+	err := invalidL7PortRule.Sanitize()
+	c.Assert(err, Not(IsNil))
```

Comment:
This test fails at the moment -- error is nil.

Reply:
The issue is with the new validation of the ports when L7 rules are present. The ports have already been validated above, so the added validation always succeeds. The validation is best moved to the beginning of `L7Rules.sanitize()` and it needs to explicitly check for port value `"0"`, like so:
```
diff --git a/pkg/policy/api/rule_validation.go b/pkg/policy/api/rule_validation.go
index 695de01559..f9fc2d07aa 100644
--- a/pkg/policy/api/rule_validation.go
+++ b/pkg/policy/api/rule_validation.go
@@ -284,6 +284,12 @@ func (e *EgressRule) sanitize() error {
 }
 
 func (pr *L7Rules) sanitize(ports []PortProtocol) error {
+	for _, pp := range ports {
+		if pp.Port == "0" {
+			return fmt.Errorf("L7 rules can not be used when a port is 0")
+		}
+	}
+
 	nTypes := 0
 
 	if pr.HTTP != nil {
@@ -362,11 +368,6 @@ func (pr *PortRule) sanitize(ingress bool) error {
 
 	// Sanitize L7 rules
 	if !pr.Rules.IsEmpty() {
-		for _, pp := range pr.Ports {
-			if err := pp.Sanitize(); err != nil {
-				return err
-			}
-		}
 		if err := pr.Rules.sanitize(pr.Ports); err != nil {
 			return err
 		}
```
With this change the new unit test passes as expected.

Reply:
@mattfenwick Sorry for the delay!

Reply:
np thanks @jrajahalme !